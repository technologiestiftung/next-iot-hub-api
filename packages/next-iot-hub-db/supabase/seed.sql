insert into users (id, name)
values (
    uuid_generate_v4 (),
    substring(
      (md5(random()::text))
      from 0 for 19
    )
  );
insert into categories ("name", description)
values ('CO2', 'foo');
insert into projects ("userId", name, "categoryId")
values (
    (
      select id
      from users
      order by random()
      limit 1
    ), md5(random()::text), (
      select id
      from categories
      order by random()
      limit 1
    )
  );
insert into devices ("userId", "externalId", "projectId")
values (
    (
      select id
      from users
      order by random()
      limit 1
    ), 'foo', (
      select id
      from projects
      order by random()
      limit 1
    )
  );
insert into records ("deviceId", "recordedAt")
values (
    (
      select id
      from devices
      order by random()
      limit 1
    ), now()
  );