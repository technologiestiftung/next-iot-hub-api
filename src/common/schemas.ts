// Copyright (c) 2021 Technologiestiftung Berlin & Fabian Mor√≥n Zirfas
//
// This software is released under the MIT License.
// https://opensource.org/licenses/MIT
import S from "fluent-json-schema";

export const getResponseDefaultSchema = S.object()
  .id("get-response-default")
  .title("default response schema for all get requests")
  .prop(
    "200",
    S.object()
      .additionalProperties(true)
      .prop("data", S.array())
      .prop("url", S.string())
      .prop("nextPage", S.string())
  );
