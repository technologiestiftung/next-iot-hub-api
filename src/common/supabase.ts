/**
 * This file was auto-generated by openapi-typescript.
 * Do not make direct changes to the file.
 */

export interface paths {
  "/": {
    get: {
      responses: {
        /** OK */
        200: unknown;
      };
    };
  };
  "/authtokens": {
    get: {
      parameters: {
        query: {
          id?: parameters["rowFilter.authtokens.id"];
          description?: parameters["rowFilter.authtokens.description"];
          projectId?: parameters["rowFilter.authtokens.projectId"];
          userId?: parameters["rowFilter.authtokens.userId"];
          niceId?: parameters["rowFilter.authtokens.niceId"];
          /** Filtering Columns */
          select?: parameters["select"];
          /** Ordering */
          order?: parameters["order"];
          /** Limiting and Pagination */
          offset?: parameters["offset"];
          /** Limiting and Pagination */
          limit?: parameters["limit"];
        };
        header: {
          /** Limiting and Pagination */
          Range?: parameters["range"];
          /** Limiting and Pagination */
          "Range-Unit"?: parameters["rangeUnit"];
          /** Preference */
          Prefer?: parameters["preferCount"];
        };
      };
      responses: {
        /** OK */
        200: {
          schema: definitions["authtokens"][];
        };
        /** Partial Content */
        206: unknown;
      };
    };
    post: {
      parameters: {
        body: {
          /** authtokens */
          authtokens?: definitions["authtokens"];
        };
        query: {
          /** Filtering Columns */
          select?: parameters["select"];
        };
        header: {
          /** Preference */
          Prefer?: parameters["preferReturn"];
        };
      };
      responses: {
        /** Created */
        201: unknown;
      };
    };
    delete: {
      parameters: {
        query: {
          id?: parameters["rowFilter.authtokens.id"];
          description?: parameters["rowFilter.authtokens.description"];
          projectId?: parameters["rowFilter.authtokens.projectId"];
          userId?: parameters["rowFilter.authtokens.userId"];
          niceId?: parameters["rowFilter.authtokens.niceId"];
        };
        header: {
          /** Preference */
          Prefer?: parameters["preferReturn"];
        };
      };
      responses: {
        /** No Content */
        204: never;
      };
    };
    patch: {
      parameters: {
        query: {
          id?: parameters["rowFilter.authtokens.id"];
          description?: parameters["rowFilter.authtokens.description"];
          projectId?: parameters["rowFilter.authtokens.projectId"];
          userId?: parameters["rowFilter.authtokens.userId"];
          niceId?: parameters["rowFilter.authtokens.niceId"];
        };
        body: {
          /** authtokens */
          authtokens?: definitions["authtokens"];
        };
        header: {
          /** Preference */
          Prefer?: parameters["preferReturn"];
        };
      };
      responses: {
        /** No Content */
        204: never;
      };
    };
  };
  "/categories": {
    get: {
      parameters: {
        query: {
          id?: parameters["rowFilter.categories.id"];
          name?: parameters["rowFilter.categories.name"];
          description?: parameters["rowFilter.categories.description"];
          /** Filtering Columns */
          select?: parameters["select"];
          /** Ordering */
          order?: parameters["order"];
          /** Limiting and Pagination */
          offset?: parameters["offset"];
          /** Limiting and Pagination */
          limit?: parameters["limit"];
        };
        header: {
          /** Limiting and Pagination */
          Range?: parameters["range"];
          /** Limiting and Pagination */
          "Range-Unit"?: parameters["rangeUnit"];
          /** Preference */
          Prefer?: parameters["preferCount"];
        };
      };
      responses: {
        /** OK */
        200: {
          schema: definitions["categories"][];
        };
        /** Partial Content */
        206: unknown;
      };
    };
    post: {
      parameters: {
        body: {
          /** categories */
          categories?: definitions["categories"];
        };
        query: {
          /** Filtering Columns */
          select?: parameters["select"];
        };
        header: {
          /** Preference */
          Prefer?: parameters["preferReturn"];
        };
      };
      responses: {
        /** Created */
        201: unknown;
      };
    };
    delete: {
      parameters: {
        query: {
          id?: parameters["rowFilter.categories.id"];
          name?: parameters["rowFilter.categories.name"];
          description?: parameters["rowFilter.categories.description"];
        };
        header: {
          /** Preference */
          Prefer?: parameters["preferReturn"];
        };
      };
      responses: {
        /** No Content */
        204: never;
      };
    };
    patch: {
      parameters: {
        query: {
          id?: parameters["rowFilter.categories.id"];
          name?: parameters["rowFilter.categories.name"];
          description?: parameters["rowFilter.categories.description"];
        };
        body: {
          /** categories */
          categories?: definitions["categories"];
        };
        header: {
          /** Preference */
          Prefer?: parameters["preferReturn"];
        };
      };
      responses: {
        /** No Content */
        204: never;
      };
    };
  };
  "/devices": {
    get: {
      parameters: {
        query: {
          id?: parameters["rowFilter.devices.id"];
          externalId?: parameters["rowFilter.devices.externalId"];
          name?: parameters["rowFilter.devices.name"];
          projectId?: parameters["rowFilter.devices.projectId"];
          userId?: parameters["rowFilter.devices.userId"];
          /** Filtering Columns */
          select?: parameters["select"];
          /** Ordering */
          order?: parameters["order"];
          /** Limiting and Pagination */
          offset?: parameters["offset"];
          /** Limiting and Pagination */
          limit?: parameters["limit"];
        };
        header: {
          /** Limiting and Pagination */
          Range?: parameters["range"];
          /** Limiting and Pagination */
          "Range-Unit"?: parameters["rangeUnit"];
          /** Preference */
          Prefer?: parameters["preferCount"];
        };
      };
      responses: {
        /** OK */
        200: {
          schema: definitions["devices"][];
        };
        /** Partial Content */
        206: unknown;
      };
    };
    post: {
      parameters: {
        body: {
          /** devices */
          devices?: definitions["devices"];
        };
        query: {
          /** Filtering Columns */
          select?: parameters["select"];
        };
        header: {
          /** Preference */
          Prefer?: parameters["preferReturn"];
        };
      };
      responses: {
        /** Created */
        201: unknown;
      };
    };
    delete: {
      parameters: {
        query: {
          id?: parameters["rowFilter.devices.id"];
          externalId?: parameters["rowFilter.devices.externalId"];
          name?: parameters["rowFilter.devices.name"];
          projectId?: parameters["rowFilter.devices.projectId"];
          userId?: parameters["rowFilter.devices.userId"];
        };
        header: {
          /** Preference */
          Prefer?: parameters["preferReturn"];
        };
      };
      responses: {
        /** No Content */
        204: never;
      };
    };
    patch: {
      parameters: {
        query: {
          id?: parameters["rowFilter.devices.id"];
          externalId?: parameters["rowFilter.devices.externalId"];
          name?: parameters["rowFilter.devices.name"];
          projectId?: parameters["rowFilter.devices.projectId"];
          userId?: parameters["rowFilter.devices.userId"];
        };
        body: {
          /** devices */
          devices?: definitions["devices"];
        };
        header: {
          /** Preference */
          Prefer?: parameters["preferReturn"];
        };
      };
      responses: {
        /** No Content */
        204: never;
      };
    };
  };
  "/projects": {
    get: {
      parameters: {
        query: {
          id?: parameters["rowFilter.projects.id"];
          name?: parameters["rowFilter.projects.name"];
          description?: parameters["rowFilter.projects.description"];
          createdAt?: parameters["rowFilter.projects.createdAt"];
          connectype?: parameters["rowFilter.projects.connectype"];
          location?: parameters["rowFilter.projects.location"];
          userId?: parameters["rowFilter.projects.userId"];
          categoryId?: parameters["rowFilter.projects.categoryId"];
          /** Filtering Columns */
          select?: parameters["select"];
          /** Ordering */
          order?: parameters["order"];
          /** Limiting and Pagination */
          offset?: parameters["offset"];
          /** Limiting and Pagination */
          limit?: parameters["limit"];
        };
        header: {
          /** Limiting and Pagination */
          Range?: parameters["range"];
          /** Limiting and Pagination */
          "Range-Unit"?: parameters["rangeUnit"];
          /** Preference */
          Prefer?: parameters["preferCount"];
        };
      };
      responses: {
        /** OK */
        200: {
          schema: definitions["projects"][];
        };
        /** Partial Content */
        206: unknown;
      };
    };
    post: {
      parameters: {
        body: {
          /** projects */
          projects?: definitions["projects"];
        };
        query: {
          /** Filtering Columns */
          select?: parameters["select"];
        };
        header: {
          /** Preference */
          Prefer?: parameters["preferReturn"];
        };
      };
      responses: {
        /** Created */
        201: unknown;
      };
    };
    delete: {
      parameters: {
        query: {
          id?: parameters["rowFilter.projects.id"];
          name?: parameters["rowFilter.projects.name"];
          description?: parameters["rowFilter.projects.description"];
          createdAt?: parameters["rowFilter.projects.createdAt"];
          connectype?: parameters["rowFilter.projects.connectype"];
          location?: parameters["rowFilter.projects.location"];
          userId?: parameters["rowFilter.projects.userId"];
          categoryId?: parameters["rowFilter.projects.categoryId"];
        };
        header: {
          /** Preference */
          Prefer?: parameters["preferReturn"];
        };
      };
      responses: {
        /** No Content */
        204: never;
      };
    };
    patch: {
      parameters: {
        query: {
          id?: parameters["rowFilter.projects.id"];
          name?: parameters["rowFilter.projects.name"];
          description?: parameters["rowFilter.projects.description"];
          createdAt?: parameters["rowFilter.projects.createdAt"];
          connectype?: parameters["rowFilter.projects.connectype"];
          location?: parameters["rowFilter.projects.location"];
          userId?: parameters["rowFilter.projects.userId"];
          categoryId?: parameters["rowFilter.projects.categoryId"];
        };
        body: {
          /** projects */
          projects?: definitions["projects"];
        };
        header: {
          /** Preference */
          Prefer?: parameters["preferReturn"];
        };
      };
      responses: {
        /** No Content */
        204: never;
      };
    };
  };
  "/records": {
    get: {
      parameters: {
        query: {
          id?: parameters["rowFilter.records.id"];
          recordedAt?: parameters["rowFilter.records.recordedAt"];
          measurements?: parameters["rowFilter.records.measurements"];
          longitude?: parameters["rowFilter.records.longitude"];
          latitude?: parameters["rowFilter.records.latitude"];
          altitude?: parameters["rowFilter.records.altitude"];
          deviceId?: parameters["rowFilter.records.deviceId"];
          /** Filtering Columns */
          select?: parameters["select"];
          /** Ordering */
          order?: parameters["order"];
          /** Limiting and Pagination */
          offset?: parameters["offset"];
          /** Limiting and Pagination */
          limit?: parameters["limit"];
        };
        header: {
          /** Limiting and Pagination */
          Range?: parameters["range"];
          /** Limiting and Pagination */
          "Range-Unit"?: parameters["rangeUnit"];
          /** Preference */
          Prefer?: parameters["preferCount"];
        };
      };
      responses: {
        /** OK */
        200: {
          schema: definitions["records"][];
        };
        /** Partial Content */
        206: unknown;
      };
    };
    post: {
      parameters: {
        body: {
          /** records */
          records?: definitions["records"];
        };
        query: {
          /** Filtering Columns */
          select?: parameters["select"];
        };
        header: {
          /** Preference */
          Prefer?: parameters["preferReturn"];
        };
      };
      responses: {
        /** Created */
        201: unknown;
      };
    };
    delete: {
      parameters: {
        query: {
          id?: parameters["rowFilter.records.id"];
          recordedAt?: parameters["rowFilter.records.recordedAt"];
          measurements?: parameters["rowFilter.records.measurements"];
          longitude?: parameters["rowFilter.records.longitude"];
          latitude?: parameters["rowFilter.records.latitude"];
          altitude?: parameters["rowFilter.records.altitude"];
          deviceId?: parameters["rowFilter.records.deviceId"];
        };
        header: {
          /** Preference */
          Prefer?: parameters["preferReturn"];
        };
      };
      responses: {
        /** No Content */
        204: never;
      };
    };
    patch: {
      parameters: {
        query: {
          id?: parameters["rowFilter.records.id"];
          recordedAt?: parameters["rowFilter.records.recordedAt"];
          measurements?: parameters["rowFilter.records.measurements"];
          longitude?: parameters["rowFilter.records.longitude"];
          latitude?: parameters["rowFilter.records.latitude"];
          altitude?: parameters["rowFilter.records.altitude"];
          deviceId?: parameters["rowFilter.records.deviceId"];
        };
        body: {
          /** records */
          records?: definitions["records"];
        };
        header: {
          /** Preference */
          Prefer?: parameters["preferReturn"];
        };
      };
      responses: {
        /** No Content */
        204: never;
      };
    };
  };
  "/users": {
    get: {
      parameters: {
        query: {
          id?: parameters["rowFilter.users.id"];
          name?: parameters["rowFilter.users.name"];
          createdAt?: parameters["rowFilter.users.createdAt"];
          role?: parameters["rowFilter.users.role"];
          /** Filtering Columns */
          select?: parameters["select"];
          /** Ordering */
          order?: parameters["order"];
          /** Limiting and Pagination */
          offset?: parameters["offset"];
          /** Limiting and Pagination */
          limit?: parameters["limit"];
        };
        header: {
          /** Limiting and Pagination */
          Range?: parameters["range"];
          /** Limiting and Pagination */
          "Range-Unit"?: parameters["rangeUnit"];
          /** Preference */
          Prefer?: parameters["preferCount"];
        };
      };
      responses: {
        /** OK */
        200: {
          schema: definitions["users"][];
        };
        /** Partial Content */
        206: unknown;
      };
    };
    post: {
      parameters: {
        body: {
          /** users */
          users?: definitions["users"];
        };
        query: {
          /** Filtering Columns */
          select?: parameters["select"];
        };
        header: {
          /** Preference */
          Prefer?: parameters["preferReturn"];
        };
      };
      responses: {
        /** Created */
        201: unknown;
      };
    };
    delete: {
      parameters: {
        query: {
          id?: parameters["rowFilter.users.id"];
          name?: parameters["rowFilter.users.name"];
          createdAt?: parameters["rowFilter.users.createdAt"];
          role?: parameters["rowFilter.users.role"];
        };
        header: {
          /** Preference */
          Prefer?: parameters["preferReturn"];
        };
      };
      responses: {
        /** No Content */
        204: never;
      };
    };
    patch: {
      parameters: {
        query: {
          id?: parameters["rowFilter.users.id"];
          name?: parameters["rowFilter.users.name"];
          createdAt?: parameters["rowFilter.users.createdAt"];
          role?: parameters["rowFilter.users.role"];
        };
        body: {
          /** users */
          users?: definitions["users"];
        };
        header: {
          /** Preference */
          Prefer?: parameters["preferReturn"];
        };
      };
      responses: {
        /** No Content */
        204: never;
      };
    };
  };
  "/rpc/delete_user": {
    post: {
      parameters: {
        body: {
          args: { [key: string]: any };
        };
        header: {
          /** Preference */
          Prefer?: parameters["preferParams"];
        };
      };
      responses: {
        /** OK */
        200: unknown;
      };
    };
  };
  "/rpc/handle_new_user": {
    post: {
      parameters: {
        body: {
          args: { [key: string]: any };
        };
        header: {
          /** Preference */
          Prefer?: parameters["preferParams"];
        };
      };
      responses: {
        /** OK */
        200: unknown;
      };
    };
  };
}

export interface definitions {
  authtokens: {
    /**
     * Note:
     * This is a Primary Key.<pk/>
     */
    id: string;
    description: string;
    /**
     * Note:
     * This is a Foreign Key to `projects.id`.<fk table='projects' column='id'/>
     */
    projectId: number;
    /**
     * Note:
     * This is a Foreign Key to `users.id`.<fk table='users' column='id'/>
     */
    userId?: string;
    niceId: number;
  };
  categories: {
    /**
     * Note:
     * This is a Primary Key.<pk/>
     */
    id: number;
    name:
      | "Temperatur"
      | "CO2"
      | "Luftfeuchtigkeit"
      | "Druck"
      | "PAXCounter"
      | "Lautstärke";
    description: string;
  };
  devices: {
    /**
     * Note:
     * This is a Primary Key.<pk/>
     */
    id: number;
    externalId: string;
    name?: string;
    /**
     * Note:
     * This is a Foreign Key to `projects.id`.<fk table='projects' column='id'/>
     */
    projectId: number;
    /**
     * Note:
     * This is a Foreign Key to `users.id`.<fk table='users' column='id'/>
     */
    userId?: string;
  };
  projects: {
    /**
     * Note:
     * This is a Primary Key.<pk/>
     */
    id: number;
    name: string;
    description?: string;
    createdAt: string;
    connectype: "ttn" | "other";
    location?: string;
    /**
     * Note:
     * This is a Foreign Key to `users.id`.<fk table='users' column='id'/>
     */
    userId: string;
    /**
     * Note:
     * This is a Foreign Key to `categories.id`.<fk table='categories' column='id'/>
     */
    categoryId: number;
  };
  records: {
    /**
     * Note:
     * This is a Primary Key.<pk/>
     */
    id: number;
    recordedAt: string;
    measurements?: string;
    longitude?: number;
    latitude?: number;
    altitude?: number;
    /**
     * Note:
     * This is a Foreign Key to `devices.id`.<fk table='devices' column='id'/>
     */
    deviceId: number;
  };
  users: {
    /**
     * Note:
     * This is a Primary Key.<pk/>
     */
    id: string;
    name?: string;
    createdAt: string;
    role?: "maker";
  };
}

export interface parameters {
  /** Preference */
  preferParams: "params=single-object";
  /** Preference */
  preferReturn: "return=representation" | "return=minimal" | "return=none";
  /** Preference */
  preferCount: "count=none";
  /** Filtering Columns */
  select: string;
  /** On Conflict */
  on_conflict: string;
  /** Ordering */
  order: string;
  /** Limiting and Pagination */
  range: string;
  /** Limiting and Pagination */
  rangeUnit: string;
  /** Limiting and Pagination */
  offset: string;
  /** Limiting and Pagination */
  limit: string;
  /** authtokens */
  "body.authtokens": definitions["authtokens"];
  "rowFilter.authtokens.id": string;
  "rowFilter.authtokens.description": string;
  "rowFilter.authtokens.projectId": string;
  "rowFilter.authtokens.userId": string;
  "rowFilter.authtokens.niceId": string;
  /** categories */
  "body.categories": definitions["categories"];
  "rowFilter.categories.id": string;
  "rowFilter.categories.name": string;
  "rowFilter.categories.description": string;
  /** devices */
  "body.devices": definitions["devices"];
  "rowFilter.devices.id": string;
  "rowFilter.devices.externalId": string;
  "rowFilter.devices.name": string;
  "rowFilter.devices.projectId": string;
  "rowFilter.devices.userId": string;
  /** projects */
  "body.projects": definitions["projects"];
  "rowFilter.projects.id": string;
  "rowFilter.projects.name": string;
  "rowFilter.projects.description": string;
  "rowFilter.projects.createdAt": string;
  "rowFilter.projects.connectype": string;
  "rowFilter.projects.location": string;
  "rowFilter.projects.userId": string;
  "rowFilter.projects.categoryId": string;
  /** records */
  "body.records": definitions["records"];
  "rowFilter.records.id": string;
  "rowFilter.records.recordedAt": string;
  "rowFilter.records.measurements": string;
  "rowFilter.records.longitude": string;
  "rowFilter.records.latitude": string;
  "rowFilter.records.altitude": string;
  "rowFilter.records.deviceId": string;
  /** users */
  "body.users": definitions["users"];
  "rowFilter.users.id": string;
  "rowFilter.users.name": string;
  "rowFilter.users.createdAt": string;
  "rowFilter.users.role": string;
}

export interface operations {}
