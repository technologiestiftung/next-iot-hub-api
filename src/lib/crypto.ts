import crypto from "crypto";
import util from "util";
const pbkdf2 = util.promisify(crypto.pbkdf2);
const randomBytes = util.promisify(crypto.randomBytes);

/**
 *
 * @param options
 * @returns
 */
export async function hash({
  token,
  iterations = 10000,
  keylen = 64,
  digest = "RSA-SHA256",
  salt,
}: {
  token: string;
  iterations?: number;
  keylen?: number;
  digest?: string;
  salt?: string;
}): Promise<{
  computedHash: string;
  salt: string;
  iterations: number;
  keylen: number;
  digest: string;
}> {
  salt = salt !== undefined ? salt : (await randomBytes(16)).toString("base64");
  const computedHash = (
    await pbkdf2(token, salt, iterations, keylen, digest)
  ).toString("base64");
  return { computedHash, salt, iterations, keylen, digest };
}

/**
 * Function to verify a token hash against the original.
 * Needs the salt used to generate the token
 * @example
 * const { hash, salt } = await hashToken({ token: "123" });
 * const result = await compare({ provided: "123", stored: hash, salt });
 */
export async function compare({
  provided,
  stored,
  salt,
}: {
  provided: string;
  stored: string;
  salt: string;
}): Promise<boolean> {
  const { computedHash } = await hash({ token: provided, salt });
  if (computedHash === stored) {
    return true;
  }
  return false;
}
