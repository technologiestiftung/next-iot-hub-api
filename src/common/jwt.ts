import { definitions } from "./supabase";
type Scope = definitions["auth_tokens"]["scope"];

export interface AuthToken {
  sub: string;
  description: string;
  jti: string;
  iss: string;
  scope: Scope;
  iat: number;
}
