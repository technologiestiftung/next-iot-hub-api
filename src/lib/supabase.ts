import fp from "fastify-plugin";
import { FastifyInstance } from "fastify";

import { createClient, SupabaseClient } from "@supabase/supabase-js";
import { definitions } from "../common/supabase";
import { compare } from "./crypto";
import * as bycrypt from "bcrypt";
type CheckAuthtokenExistsType = (
  token: string,
  userId: string
) => Promise<boolean>;
declare module "fastify" {
  export interface FastifyInstance {
    supabase: SupabaseClient;
    checkAuthtokenExists: CheckAuthtokenExistsType;
  }
}

export interface SupabasePluginOptions {
  supabaseUrl: string;
  supabaseServiceRoleKey: string;
}

const FastifySupabase = async (
  fastify: FastifyInstance,
  opts: SupabasePluginOptions
) => {
  const checkAuthtokenExists: CheckAuthtokenExistsType = async (
    token,
    userId
  ) => {
    // get all tokens from user
    const { data: authtokens, error } = await fastify.supabase
      .from<definitions["auth_tokens"]>("auth_tokens")
      .select("*")
      .eq("user_id", userId);
    if (error) {
      fastify.log.error("postgres error");
      throw fastify.httpErrors.internalServerError(error.hint);
    }
    if (!authtokens || authtokens.length === 0) {
      fastify.log.warn("no token found");
      throw fastify.httpErrors.unauthorized();
    }
    // compare all tokens ids with actual token
    let authtoken: definitions["auth_tokens"] | null = null;
    let foundOldToken = false;
    for (const at of authtokens) {
      if (at.salt.length === 0) {
        // we have one of our old tokens and we actually cant
        // verify which is which. So we at least checked that the
        // user exisits.
        const comparedOld = await bycrypt.compare(token, at.id);
        if (comparedOld) {
          foundOldToken = true;
          fastify.log.warn(
            `old token was used, token nice id ${at.nice_id} user ${userId}`
          );
        }
      } else {
        const compared = await compare({
          provided: token,
          stored: at.id,
          salt: at.salt,
        });
        if (compared === true) {
          authtoken = at;
          break;
        }
      }
    }

    if (foundOldToken) {
      return true;
    }
    return authtoken !== null ? true : false;
  };

  const supabaseAdmin = createClient(
    opts.supabaseUrl,
    opts.supabaseServiceRoleKey,
    {
      autoRefreshToken: false,
      persistSession: false,
      detectSessionInUrl: false,
      realtime: {
        timeout: 1,
      },
    }
  );
  if (!fastify.supabase) {
    fastify.decorate("supabase", supabaseAdmin);
    fastify.decorate("checkAuthtokenExists", checkAuthtokenExists);
  }
};
export default fp(FastifySupabase);
