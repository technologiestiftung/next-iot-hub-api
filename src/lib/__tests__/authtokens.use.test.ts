/* eslint-disable @typescript-eslint/no-non-null-assertion */
// Copyright (c) 2021 Technologiestiftung Berlin & Fabian MorÃ³n Zirfas
//
// This software is released under the MIT License.
// https://opensource.org/licenses/MIT

import { definitions } from "../../common/supabase";
import {
  deleteUser,
  supabaseAnonKey,
  apiVersion,
  buildServerOpts,
  truncateTables,
  signupUser,
  closePool,
  connectPool,
  createSensor,
  createTTNPayload,
} from "../../__test-utils";
import { compare } from "../crypto";
import buildServer from "../server";

describe("authtokens POST tests", () => {
  beforeAll(async () => {
    await connectPool();
  });
  beforeEach(async () => {
    await truncateTables();
  });
  afterAll(async () => {
    // await truncateTables();
    await closePool();
  });

  test("should be able to identify tokens by their hash", async () => {
    const server = buildServer(buildServerOpts);
    const user = await signupUser();

    const url = `/api/v${apiVersion}/authtokens`;
    const response1 = await server.inject({
      method: "POST",
      url,
      headers: {
        authorization: `Bearer ${user.token}`,
        apikey: supabaseAnonKey,
      },
      payload: {
        description: "testing token usage",
      },
    });

    const response2 = await server.inject({
      method: "POST",
      url,
      headers: {
        authorization: `Bearer ${user.token}`,
        apikey: supabaseAnonKey,
      },
      payload: {
        description: "testing token usage",
      },
    });

    const { token: token1 } = response1.json<{
      data: { token: string; nice_id: number };
    }>().data;

    const { token: token2 } = response2.json<{
      data: { token: string; nice_id: number };
    }>().data;

    const { data: tokens, error } = await server.supabase
      .from<definitions["auth_tokens"]>("auth_tokens")
      .select("*")
      .eq("user_id", user.id);
    expect(response1.statusCode).toBe(201);
    expect(JSON.parse(response1.body)).toMatchSnapshot({
      data: { token: expect.any(String), nice_id: expect.any(Number) },
    });
    expect(response2.statusCode).toBe(201);
    expect(JSON.parse(response2.body)).toMatchSnapshot({
      data: { token: expect.any(String), nice_id: expect.any(Number) },
    });

    expect(tokens).not.toBeNull();
    expect(error).toBeNull();
    expect(tokens!).toHaveLength(2);

    const actual1 = await compare({
      provided: token1,
      stored: tokens![0].id,
      salt: tokens![0].salt,
    });

    expect(actual1).toBeTruthy();

    const actual2 = await compare({
      provided: token2,
      stored: tokens![1].id,
      salt: tokens![1].salt,
    });

    expect(actual2).toBeTruthy();
    await deleteUser(user.token);
  });

  test("should be able to have two tokens and use them both", async () => {
    const server = buildServer(buildServerOpts);
    const user = await signupUser();
    const httpPayload = {
      latitude: 52.483107,
      longitude: 13.390679,
      altitude: 30,
      measurements: [1, 2, 3],
    };
    const payload = createTTNPayload({
      end_device_ids: {
        device_id: "test-device-id",
        application_ids: {
          application_id: "test-application-id",
        },
      },

      uplink_message: {
        decoded_payload: {
          measurements: [1, 2, 3],
        },
        locations: {
          user: {
            latitude: 1,
          },
        },
      },
    });
    const httpSensor = await createSensor({
      user_id: user.id,
      name: "test with multiple tokens",
    });
    const ttnSensor = await createSensor({
      user_id: user.id,
      name: "test with multiple tokens",
      external_id: payload.end_device_ids.device_id,
    }); //?

    const url = `/api/v${apiVersion}/authtokens`;
    const response1 = await server.inject({
      method: "POST",
      url,
      headers: {
        authorization: `Bearer ${user.token}`,
        apikey: supabaseAnonKey,
      },
      payload: {
        description: "testing token usage",
      },
    });

    const response2 = await server.inject({
      method: "POST",
      url,
      headers: {
        authorization: `Bearer ${user.token}`,
        apikey: supabaseAnonKey,
      },
      payload: {
        description: "testing token usage",
      },
    });

    const { token: token1 } = response1.json<{
      data: { token: string; nice_id: number };
    }>().data;

    const { token: token2 } = response2.json<{
      data: { token: string; nice_id: number };
    }>().data;

    const response3 = await server.inject({
      method: "POST",
      url: `/api/v${apiVersion}/sensors/${httpSensor.id}/records`,
      payload: httpPayload,
      headers: {
        Authorization: `Bearer ${token1}`,
      },
    });
    const response4 = await server.inject({
      method: "POST",
      url: `/api/v${apiVersion}/sensors/${httpSensor.id}/records`,
      payload: httpPayload,
      headers: {
        Authorization: `Bearer ${token2}`,
      },
    });
    const response5 = await server.inject({
      method: "POST",
      url: `/api/v${apiVersion}/integrations/ttn/v3`,
      payload,
      headers: {
        Authorization: `Bearer ${token1}`,
      },
    });
    const response6 = await server.inject({
      method: "POST",
      url: `/api/v${apiVersion}/integrations/ttn/v3`,
      payload,
      headers: {
        Authorization: `Bearer ${token2}`,
      },
    });
    const { data: httpRecords, error: httpError } = await server.supabase
      .from<definitions["records"]>("records")
      .select("*")
      .eq("sensor_id", httpSensor.id);
    const { data: ttnRecords, error: ttnError } = await server.supabase
      .from<definitions["records"]>("records")
      .select("*")
      .eq("sensor_id", httpSensor.id);

    expect(response3.statusCode).toBe(201);
    expect(response4.statusCode).toBe(201);
    expect(response5.statusCode).toBe(201);
    expect(response6.statusCode).toBe(201);
    expect(httpRecords).not.toBeNull();
    expect(ttnRecords).not.toBeNull();
    expect(httpError).toBeNull();
    expect(ttnError).toBeNull();
    expect(httpRecords!).toHaveLength(2);
    expect(ttnRecords!).toHaveLength(2);
    // start boilerplate delete user
    // await deleteUser(user.token);
    // end boilerplate
  });
});
