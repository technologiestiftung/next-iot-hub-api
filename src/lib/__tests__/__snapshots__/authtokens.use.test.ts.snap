// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`authtokens POST tests should be able to identify tokens by their hash 1`] = `
Object {
  "data": Object {
    "nice_id": Any<Number>,
    "token": Any<String>,
  },
  "method": "POST",
  "url": "/api/v3/authtokens",
}
`;

exports[`authtokens POST tests should be able to identify tokens by their hash 2`] = `
Object {
  "data": Object {
    "nice_id": Any<Number>,
    "token": Any<String>,
  },
  "method": "POST",
  "url": "/api/v3/authtokens",
}
`;
